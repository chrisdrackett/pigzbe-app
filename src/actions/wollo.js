import Stellar from '../stellar';

const server = new Stellar.Server('https://horizon-testnet.stellar.org');
Stellar.Network.useTestNetwork();

export const UPDATE_ACCOUNT = 'UPDATE_ACCOUNT';
export const UPDATE_BALANCE = 'UPDATE_BALANCE';

const getWolloBalance = account => {
    const wollo = account.balances.find(b => b.asset_code === 'WOL');
    return wollo ? wollo.balance : '0';
};

export const updateBalance = balance => ({type: UPDATE_BALANCE, balance});

export const loadAccount = publicKey => dispatch => {
    return server.loadAccount(publicKey)
        .then(account => {
            dispatch({type: UPDATE_ACCOUNT, account});
            dispatch(updateBalance(getWolloBalance(account)));
            return account;
        });
};
