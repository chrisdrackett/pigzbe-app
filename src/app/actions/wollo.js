import {
    loadAccount,
    setServer,
    isValidPublicKey,
    paymentHistory,
    paymentInfo,
    sendPayment,
    getBalance,
    getMinBalance,
    checkHasGas,
    checkAssetTrusted,
    trustAsset,
    createAccount,
    TransactionBuilder,
    trustAssetTransaction,
    submitTransaction,
    Keypair
} from '@pigzbe/stellar-utils';
import {
    strings,
    ASSET_CODE
} from '../constants';
import Keychain from '../utils/keychain';
import {wolloAsset} from '../selectors';
import {createKeypair, appError, updateKidBalance} from './';

export const WOLLO_LOADING = 'WOLLO_LOADING';
export const WOLLO_ERROR = 'WOLLO_ERROR';
export const WOLLO_UPDATE_ACCOUNT = 'WOLLO_UPDATE_ACCOUNT';
export const WOLLO_UPDATE_BALANCE = 'WOLLO_UPDATE_BALANCE';
export const WOLLO_UPDATE_XLM = 'WOLLO_UPDATE_XLM';
export const WOLLO_UPDATE_PAYMENTS = 'WOLLO_UPDATE_PAYMENTS';
export const WOLLO_SENDING = 'WOLLO_SENDING';
export const WOLLO_SEND_COMPLETE = 'WOLLO_SEND_COMPLETE';
export const WOLLO_SEND_STATUS = 'WOLLO_SEND_STATUS';
export const WOLLO_SEND_RESET = 'WOLLO_SEND_RESET';
export const WOLLO_KEYPAIR = 'WOLLO_KEYPAIR';
export const WOLLO_KEYPAIR_SAVED = 'WOLLO_KEYPAIR_SAVED';

export const getWolloBalance = account => getBalance(account, ASSET_CODE);

export const wolloLoading = loading => ({type: WOLLO_LOADING, loading});
export const wolloSending = sending => ({type: WOLLO_SENDING, sending});
export const wolloSendComplete = sendComplete => ({type: WOLLO_SEND_COMPLETE, sendComplete});
export const wolloSendStatus = sendStatus => ({type: WOLLO_SEND_STATUS, sendStatus});
export const wolloSendReset = () => ({type: WOLLO_SEND_RESET});

export const wolloError = error => ({type: WOLLO_ERROR, error});

export const setHorizonURI = uri => () => {
    console.log('setHorizonURI:', uri);
    setServer(uri);
};

const updateBalance = balance => ({type: WOLLO_UPDATE_BALANCE, balance});

const updateXLM = account => dispatch => {
    const balanceXLM = getBalance(account);
    const minXLM = getMinBalance(account, 10000);
    console.log('minXLM', minXLM);
    const hasGas = checkHasGas(account);
    dispatch({type: WOLLO_UPDATE_XLM, balanceXLM, minXLM, hasGas});
};

export const loadWallet = publicKey => async (dispatch, getState) => {
    console.log('loadWallet');
    const key = publicKey || getState().keys.publicKey;
    try {
        if (key) {
            const account = await loadAccount(key);
            console.log('account', account);
            dispatch({type: WOLLO_UPDATE_ACCOUNT, account});
            dispatch(updateBalance(getWolloBalance(account)));

            const asset = wolloAsset(getState());
            const isTrusted = checkAssetTrusted(account, asset);
            if (!isTrusted) {
                await trustAsset(getState().keys.secretKey, asset);
            }
            dispatch(updateXLM(account));
        }
    } catch (error) {
        console.log('Could not load wallet with publicKey', key);
    }
};

export const refreshBalance = kidAddress => async (dispatch, getState) => {
    const key = getState().keys.publicKey;
    try {
        if (key) {
            const account = await loadAccount(key);
            dispatch({type: WOLLO_UPDATE_ACCOUNT, account});
            dispatch(updateBalance(getWolloBalance(account)));
            dispatch(updateXLM(account));
        }
        if (kidAddress) {
            const account = await loadAccount(kidAddress);
            dispatch(updateKidBalance(kidAddress, getWolloBalance(account)));
        }
    } catch (error) {
        console.log('Could not load wallet with publicKey', key);
    }
};

export const loadPayments = address => async (dispatch, getState) => {
    const {publicKey} = getState().keys;
    const key = address || publicKey;

    dispatch(wolloLoading(true));
    dispatch(wolloError(null));

    try {
        const rawData = await paymentHistory(key);
        const filteredData = rawData.filter(p => p.type !== 'account_merge');
        const payments = await Promise.all(filteredData.map(p => paymentInfo(p, key)));
        dispatch({type: WOLLO_UPDATE_PAYMENTS, payments});
    } catch (error) {
        console.log(error);
    }

    dispatch(wolloLoading(false));
};


export const sendWollo = (destination, amount, memo) => async (dispatch, getState) => {
    console.log('sendWollo', destination, amount, memo);
    if (!isValidPublicKey(destination)) {
        dispatch(wolloError(strings.transferErrorInvalidKey));
        dispatch(appError(strings.transferErrorInvalidKey));
        return;
    }

    if (!amount || isNaN(amount) || Number(amount) === 0) {
        dispatch(wolloError(strings.transferErrorInvalidAmount));
        dispatch(appError(strings.transferErrorInvalidAmount));
        return;
    }

    dispatch(wolloError(null));
    dispatch(wolloSendComplete(false));
    dispatch(wolloSending(true));
    dispatch(wolloSendStatus(strings.transferStatusChecking));

    const {secretKey, publicKey} = getState().keys;

    let destAccount;

    try {
        destAccount = await loadAccount(destination);
    } catch (e) {
        console.log(e);
    }

    if (!destAccount) {
        dispatch(wolloSending(false));
        dispatch(wolloSendStatus(null));
        dispatch(wolloError(strings.transferStatusInvalidDestination));
        dispatch(appError(strings.transferStatusInvalidDestination));
        return;
    }

    const asset = wolloAsset(getState());

    const isTrusted = checkAssetTrusted(destAccount, asset);

    if (!isTrusted) {
        dispatch(wolloSending(false));
        dispatch(wolloSendStatus(null));
        dispatch(wolloError(strings.transferErrorTrust));
        dispatch(appError(strings.transferErrorTrust));
        return;
    }

    dispatch(wolloSendStatus(strings.transferStatusSending));

    let result;

    try {
        result = await sendPayment(secretKey, destination, amount, memo, asset);
    } catch (e) {
        console.error(e);
    }

    if (!result) {
        dispatch(wolloSending(false));
        dispatch(wolloSendStatus(null));
        dispatch(wolloError(strings.transferStatusFailed));
        dispatch(appError(strings.transferStatusFailed));
        return;
    }

    dispatch(refreshBalance());
    dispatch(wolloSending(false));
    dispatch(wolloSendComplete(true));
    dispatch(wolloSendStatus(strings.transferStatusComplete));
    dispatch(wolloError(null));
    dispatch(loadWallet(publicKey));
};

export const fundKidAccount = (memo, address, startingBalance) => async (dispatch, getState) => {
    console.log('fundKidAccount', memo, address, startingBalance);
    // const {publicKey, secretKey} = getState().keys;
    const {secretKey} = getState().keys;
    const kidSecretKey = await Keychain.load(`secret_${address}`);
    try {
        const keypair = Keypair.fromSecret(kidSecretKey);
        const destination = keypair.publicKey();

        const account = await createAccount(secretKey, destination, startingBalance, memo);
        console.log('account', account);

        const txb = new TransactionBuilder(account);

        const asset = wolloAsset(getState());
        trustAssetTransaction(txb, asset);

        const transaction = txb.build();
        transaction.sign(keypair);
        const result = await submitTransaction(transaction);
        console.log('result', result);
    } catch (error) {
        console.log(error);
    }
};

export const createKidAccount = (memo, nickname, startingBalance) => async dispatch => {
    console.log('createKidAccount', nickname);

    try {
        const keypair = await dispatch(createKeypair());
        const destination = keypair.publicKey();
        console.log('createKidAccount destination', destination);

        await Keychain.save(`secret_${destination}`, keypair.secret());

        console.log('createKidAccount startingBalance', startingBalance);

        await dispatch(fundKidAccount(`${memo}${nickname}`, destination, startingBalance));

        return destination;

    } catch (error) {
        console.log(error);
    }

    return null;
};
