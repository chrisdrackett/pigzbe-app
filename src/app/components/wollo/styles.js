import {StyleSheet} from 'react-native';
import {
    color,
    fontFamily
} from '../../styles';

export default StyleSheet.create({
    wolloContainer: {
        justifyContent: 'center',
        alignItems: 'center',
    },
    balanceContainer: {
        justifyContent: 'center',
        alignItems: 'center',
        flexDirection: 'row'
    },
    currencyLogo: {
        width: 23,
        height: 23,
        marginRight: 10,
    },
    balance: {
        fontFamily,
        fontWeight: 'bold',
        color: color.white,
        fontSize: 22,
    },
    label: {
        fontFamily,
        fontWeight: 'bold',
        color: color.whiteOpacity60,
        fontSize: 14,
    },
    balance__dark: {
        color: color.blue,
    },
    label__dark: {
        color: color.blue,
    },
});
