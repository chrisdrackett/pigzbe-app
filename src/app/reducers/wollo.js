import {USE_TESTNET, BASE_CURRENCY} from '../constants';

import {
    WOLLO_SET_LOADING,
    WOLLO_SET_ERROR,
    WOLLO_USE_TESTNET,
    WOLLO_UPDATE_ACCOUNT,
    WOLLO_UPDATE_BALANCE,
    WOLLO_UPDATE_PAYMENTS,
    WOLLO_SET_SENDING,
    WOLLO_UPDATE_XLM,
    WOLLO_UPDATE_ISSUER
} from '../actions';

export const initialState = {
    useTestnet: USE_TESTNET,
    balance: '0',
    balanceXLM: '0',
    minXLM: '0',
    hasGas: false,
    baseCurrency: BASE_CURRENCY,
    payments: [],
    loading: false,
    error: null,
    sending: false,
};

export default (state = initialState, action) => {
    switch (action.type) {
        case WOLLO_USE_TESTNET:
            return {
                ...state,
                useTestnet: action.useTestnet
            };
        case WOLLO_UPDATE_ACCOUNT:
            return {
                ...state,
                account: action.account
            };
        case WOLLO_UPDATE_BALANCE:
            return {
                ...state,
                balance: action.balance
            };
        case WOLLO_UPDATE_PAYMENTS:
            return {
                ...state,
                payments: action.payments
            };
        case WOLLO_SET_LOADING:
            return {
                ...state,
                loading: action.loading
            };
        case WOLLO_SET_ERROR:
            return {
                ...state,
                error: action.error
            };
        case WOLLO_SET_SENDING:
            return {
                ...state,
                sending: action.sending
            };
        case WOLLO_UPDATE_XLM:
            return {
                ...state,
                balanceXLM: action.balanceXLM,
                minXLM: action.minXLM,
                hasGas: action.hasGas
            };
        case WOLLO_UPDATE_ISSUER:
            return {
                ...state,
                issuer: action.issuer
            };
        default:
            return state;
    }
};
