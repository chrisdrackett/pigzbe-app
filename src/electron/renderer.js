import {AppRegistry} from 'react-native';
// import App from '../../App';
import App from '../../dist/bundle.web.js';

AppRegistry.registerComponent('App', () => App);
AppRegistry.runApplication('App', {rootTag: document.getElementById('react-root')});
