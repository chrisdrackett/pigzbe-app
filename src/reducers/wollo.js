import {
    UPDATE_BALANCE
} from '../actions';

const initialState = {
    balance: '0'
};

export default (state = initialState, action) => {
    switch (action.type) {
        case UPDATE_BALANCE:
            return {
                ...state,
                balance: action.balance
            };
        default:
            return state;
    }
};
