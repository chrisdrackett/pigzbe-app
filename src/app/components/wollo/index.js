import React from 'react';
import {
    Text,
    View,
    Image
} from 'react-native';
import styles from './styles';
import {
    strings,
    COIN_DPS,
    ASSET_CODE
} from '../../constants';
import moneyFormat from '../../utils/money-format';

export default ({balance}) => (
    <View style={styles.wolloContainer}>
        <View style={styles.balanceContainer}>
            <Image style={styles.currencyLogo} source={require('./images/currency_logo.png')} />
            <Text style={styles.balance}>{moneyFormat(balance, COIN_DPS[ASSET_CODE])}</Text>
        </View>
        <Text style={styles.label}>{strings.walletBalance}</Text>
    </View>
);
