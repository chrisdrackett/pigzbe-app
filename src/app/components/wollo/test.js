import React from 'react';
import renderer from 'react-test-renderer';
import Wollo from './';

const props = {
    balance: '100.0000'
};

describe('Wollo', () => {
    test('renders correctly', () => {
        const tree = renderer.create(<Wollo {...props}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
