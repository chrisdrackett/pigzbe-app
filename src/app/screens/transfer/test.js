import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import Transfer from './';
import wollo from '../../reducers/wollo';
import {mockStore} from '../../../setupTests';

const props = {
};

describe('Transfer', () => {
    test('renders correctly', () => {
        const tree = renderer.create((
            <Provider store={mockStore({wollo})}>
                <Transfer {...props}/>
            </Provider>
        )).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
