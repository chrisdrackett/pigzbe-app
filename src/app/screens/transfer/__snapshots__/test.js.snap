// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transfer renders correctly 1`] = `
Array [
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "rgb(0, 50, 120)",
          "flex": 1,
          "justifyContent": "space-between",
        },
        Object {
          "width": 750,
        },
        Object {
          "display": "flex",
          "flex": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgb(21, 74, 150)",
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 242,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgb(0, 50, 120)",
          "display": "flex",
          "flexDirection": "row",
          "height": 70,
          "justifyContent": "space-between",
          "paddingTop": 20,
        }
      }
    >
      <View
        style={
          Object {
            "display": "flex",
            "justifyContent": "center",
            "marginLeft": 18,
            "marginRight": 18,
            "width": 40,
          }
        }
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/app/components/logo/images/pigzbe_logo_inapp.png",
            }
          }
          style={
            Object {
              "height": 25,
              "width": 65,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "display": "flex",
            "justifyContent": "center",
            "marginLeft": 18,
            "marginRight": 18,
            "width": 40,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginTop": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "rgb(255, 255, 255)",
              "fontFamily": "Poppins",
              "fontSize": 22,
              "fontWeight": "bold",
            },
            undefined,
          ]
        }
      >
        Transfer
      </Text>
      <Image
        source={
          Object {
            "testUri": "../../../src/app/components/step-header/images/transfer.png",
          }
        }
        style={
          Object {
            "height": 100,
            "marginBottom": -30,
            "marginTop": 8,
            "width": 200,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "flexGrow": 1,
            },
            Object {
              "bottom": 0,
            },
          ]
        }
      >
        <View
          onLayout={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "rgb(0, 50, 120)",
                "flex": 1,
                "justifyContent": "space-between",
              },
              Object {
                "width": 750,
              },
              Array [
                Object {
                  "alignSelf": "center",
                  "borderTopLeftRadius": 5,
                  "borderTopRightRadius": 5,
                  "width": "88.75%",
                },
                Object {
                  "backgroundColor": "rgb(255, 255, 255)",
                },
              ],
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flex": 1,
                },
                null,
                Object {
                  "paddingTop": 30,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "rgb(0, 50, 120)",
                    "fontFamily": "Poppins",
                    "fontSize": 16,
                    "marginBottom": 18,
                    "textAlign": "center",
                  },
                  null,
                  undefined,
                ]
              }
            >
              No transaction history
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      Object {
        "alignSelf": "center",
        "bottom": 20,
        "left": "5.625%",
        "paddingLeft": 18,
        "paddingRight": 18,
        "position": "absolute",
        "width": "88.75%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "rgba(21, 74, 150, 0.5)",
            "borderColor": "rgba(255, 255, 255, 0)",
            "borderRadius": 22.5,
            "borderWidth": 1,
            "marginBottom": 10,
            "maxWidth": 400,
            "opacity": 1,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "color": "rgb(255, 255, 255)",
                "fontFamily": "Poppins",
                "fontSize": 14,
                "fontWeight": "bold",
                "height": 45,
                "paddingLeft": 20,
                "paddingRight": 20,
                "paddingTop": 12,
                "textAlign": "center",
              },
              undefined,
              Array [
                Object {
                  "color": "rgb(255, 255, 255)",
                },
                undefined,
              ],
              undefined,
            ]
          }
        >
          Transfer Wollo
        </Text>
      </View>
    </View>
  </View>,
]
`;
